pkg/virtualfs/directory_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 The Kanister Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualfs
+
+import (
+	"context"
+
+	. "gopkg.in/check.v1"
+)
+
+const (
+	defaultPermissions = 0777
+)
+
+type DirectorySuite struct{}
+
+var _ = Suite(&DirectorySuite{})
+
+func (s *DirectorySuite) TestAddDir(c *C) {
+	sourceDir, err := NewDirectory("root")
+	c.Assert(err, IsNil)
+
+	// New directory
+	dir, err := sourceDir.AddDir("d1", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(dir.Name(), Equals, "d1")
+
+	// Duplicate directory
+	_, err = sourceDir.AddDir("d1", defaultPermissions)
+	c.Assert(err, NotNil)
+
+	// Invalid name
+	_, err = sourceDir.AddDir("/d2", defaultPermissions)
+	c.Assert(err, NotNil)
+}
+
+func (s *DirectorySuite) TestAddAllDirs(c *C) {
+	sourceDir, err := NewDirectory("root")
+	c.Assert(err, IsNil)
+
+	c.Log("Add a directory - root/d1")
+	subdir, err := sourceDir.AddAllDirs("d1/d2", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(subdir.Name(), Equals, "d2")
+	d1, err := sourceDir.Subdir("d1")
+	c.Assert(err, IsNil)
+	c.Assert(d1, NotNil)
+
+	c.Log("Add third level directory - root/d1/d2")
+	subdir, err = sourceDir.AddAllDirs("d1/d2", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(subdir.Name(), Equals, "d2")
+	d2, err := d1.Subdir("d2")
+	c.Assert(err, IsNil)
+	c.Assert(d2, NotNil)
+
+	c.Log("Add third/fourth level dirs - root/d1/d3/d4")
+	subdir, err = sourceDir.AddAllDirs("d1/d3/d4", defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(subdir.Name(), Equals, "d4")
+	d3, err := d1.Subdir("d3")
+	c.Assert(err, IsNil)
+	c.Assert(d3, NotNil)
+	d4, err := d3.Subdir("d4")
+	c.Assert(err, IsNil)
+	c.Assert(d4, NotNil)
+
+	c.Log("Fail adding directory under a file - root/f1/d6")
+	f, err := AddFileWithContent(sourceDir, "f1", []byte("test"), defaultPermissions, defaultPermissions)
+	c.Assert(err, IsNil)
+	c.Assert(f.Name(), Equals, "f1")
+	_, err = sourceDir.AddAllDirs("f1/d6", defaultPermissions)
+	c.Assert(err, NotNil)
+}
+
+func (s *DirectorySuite) TestAddFile(c *C) {
+	sourceDir, err := NewDirectory("root")
+	c.Assert(err, IsNil)
+
+	c.Log("Add file with stream source - root/f1")
+	f, err := AddFileWithStreamSource(sourceDir, "f1", "http://test-endpoint", defaultPermissions, defaultPermissions)
```

Comment:
It'd be good to have basic tests that exercise the streaming part to prevent regression.